File: src_local/agents/Support/obfuscation_logic_engine.py
"""
SISTEMA DE PERSONAS AGENTES - SUPORTE TÉCNICO
Módulo: Motor de Lógica de Ofuscação (ObfuscationLogicEngine)
Função: Resolver concatenações AST e detectar palavras perigosas ocultas.
"""
import ast
import logging

logger = logging.getLogger(__name__)

class ObfuscationLogicEngine:
    def resolve_constant(self, node):
        """Resolve recursivamente a string concatenada."""
        if isinstance(node, (ast.Constant, getattr(ast, "Str", ast.Constant))):
            val = getattr(node, "value", getattr(node, "s", ""))
            return val if isinstance(val, str) else None
        if isinstance(node, ast.BinOp) and isinstance(node.op, ast.Add):
            l, r = self.resolve_constant(node.left), self.resolve_constant(node.right)
            return l + r if l is not None and r is not None else None
        return None

    def check_dangerous_keywords(self, node, resolved, keywords):
        import time
        start_time = time.time()
        
        for kw in keywords:
            if kw in resolved and self._is_fragmented(node, kw):
                logger.warning(f"Obfuscation: Detectado na linha {node.lineno}")
                
                from src_local.utils.logging_config import log_performance
                log_performance(logger, start_time, "Telemetry: Keyword check")
                return {
                    "line": node.lineno, "evidence": "Concatenação Suspeita", 
                    "reconstruction": resolved, "keyword": kw
                }
        return None

    def _is_fragmented(self, node, kw):
        l, r = self.resolve_constant(node.left), self.resolve_constant(node.right)
        return False if (l is None or r is None or kw in l or kw in r) else True

File: src_local/agents/Support/Security/obfuscation_logic_engine.ts
import { DANGEROUS_KEYWORDS } from "./safety_definitions";
import winston from 'winston';

/**
 * 🕵️ Motor de Lógica de Ofuscação (ObfuscationLogicEngine)
 * Resolve recursivamente concatenações e detecta palavras perigosas ocultas.
 */
export class ObfuscationLogicEngine {
    private logger = winston.loggers.get('default_logger') || winston;

    /**
     * Resolve strings concatenadas a partir de fragmentos.
     * Esta é uma implementação simplificada que deve ser usada com análise AST.
     */
    resolveConstant(node: any): string | null {
        if (node.type === 'StringLiteral' || node.type === 'Literal') {
            return node.value;
        }
        if (node.type === 'BinaryExpression' && node.operator === '+') {
            const left = this.resolveConstant(node.left);
            const right = this.resolveConstant(node.right);
            if (left !== null && right !== null) {
                return left + right;
            }
        }
        return null;
    }

    /**
     * Verifica se a string resolvida contém palavras perigosas fragmentadas.
     */
    checkDangerousKeywords(line: number, resolved: string, node: any): any | null {
        for (const kw of DANGEROUS_KEYWORDS) {
            if (resolved.includes(kw) && this.isFragmented(node, kw)) {
                return {
                    line: line,
                    evidence: "Concatenação Suspeita",
                    reconstruction: resolved,
                    keyword: kw
                };
            }
        }
        return null;
    }

    private isFragmented(node: any, kw: string): boolean {
        const left = this.resolveConstant(node.left);
        const right = this.resolveConstant(node.right);

        // Se a keyword está inteira em um dos lados, não é ofuscação por fragmentação
        if (left?.includes(kw) || right?.includes(kw)) {
            return false;
        }
        return true;
    }
}
